File: Problem-2/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) {this.val = val;}
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return add(l1, l2, 0);
    }

    public ListNode add(ListNode l1, ListNode l2, int remain) {
        if (l1==null & l2==null & remain==0) {
            return null;
        }

        int val1 = (l1 != null) ? l1.val : 0;
        int val2 = (l2 != null) ? l2.val : 0;
        int sum = val1 + val2 + remain;

        ListNode currentNode = new ListNode(sum % 10);
        if (l1 != null || l2 != null) {
            currentNode.next = add((l1 != null) ? l1.next : null, (l2 != null) ? l2.next : null, sum / 10);
        }
        return currentNode;
    }   
}

File: Problem-2/Solution.cpp
using namespace std;

#include<iostream>
#include<list>
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        return add(l1, l2, 0);
    }

    ListNode* add(ListNode* node1, ListNode* node2, int addNum = 0) {
        if (!node1 && !node2 && addNum == 0){
            return nullptr;
        }

        int val1 = (node1) ? node1->val : 0;
        int val2 = (node2) ? node2->val : 0;

        int sum = val1 + val2 + addNum;

        ListNode* currentNode = new ListNode(sum % 10);
        if (node1 || node2) {
            currentNode->next = add((node1) ? node1->next:nullptr, (node2) ? node2->next:nullptr, sum/10);
        }
        return currentNode;
    }
};
